dbms/src/Debug/dbgFuncCoprocessor.cpp

```diff
@@ -339,56 +340,99 @@ void dbgFuncTiDBQueryFromNaturalDag(Context & context, const ASTs & args, DBGInv
-
-        DAGContext dag_context(dag_request);
-        dag_context.tables_regions_info = std::move(tables_regions_info);
-        dag_context.log = log;
-        context.setDAGContext(&dag_context);
-        DAGDriver driver(context, properties.start_ts, DEFAULT_UNSPECIFIED_SCHEMA_VERSION, &dag_response, true);
-        driver.execute();
-
-        auto resp_ptr = std::make_shared<tipb::SelectResponse>();
-        if (!resp_ptr->ParseFromString(res.data()))
+        try
         {
-            throw Exception("Incorrect json response data!", ErrorCodes::BAD_ARGUMENTS);
+            unequal_flag = runAndCompareDagReq(req, res, context, unequal_msg);
         }
-        else
+        catch (const Exception & e)
         {
-            unequal_flag |= (!dagRspEqual(context, *resp_ptr, dag_response, unequal_msg));
-            if (unequal_flag)
+            failed_flag = true;
+            unequal_msg = e.message();
+        }
+        catch (...)
+        {
+            failed_flag = true;
+            unequal_msg = "Unknown execution exception!";
+        }
+
+        if (unequal_flag || failed_flag)
+        {
+            failed_req_msg_vec.push_back(std::make_pair(req_id, unequal_msg));
+            if (!dag.continueWhenError())
                 break;
         }
         ++req_idx;
     }
-    // It is all right to throw exception above, dag.clean is only to make it better
     dag.clean(context);
-    if (unequal_flag)
+    if (!failed_req_msg_vec.empty())
     {
         output("Invalid");
-        throw Exception(fmt::format("{}th request results are not equal, msg: {}", req_idx, unequal_msg), ErrorCodes::LOGICAL_ERROR);
+        String merged_msg;
+        bool first = true;
+        for (auto & it : failed_req_msg_vec)
+        {
+            String msg = fmt::format("request {} failed, msg: {}", it.first, it.second);
+            if (first)
+            {
+                merged_msg = msg;
+                first = false;
+            }
+            else
+            {
+                merged_msg = fmt::format("{}\n{}", merged_msg, msg);
+            }
+        }
+        throw Exception(merged_msg, ErrorCodes::LOGICAL_ERROR);
+    }
+}
+
+bool runAndCompareDagReq(const coprocessor::Request & req, const coprocessor::Response & res, Context & context, String & unequal_msg)
+{
+    kvrpcpb::Context req_context = req.context();
+    RegionID region_id = req_context.region_id();
+    tipb::DAGRequest dag_request = getDAGRequestFromStringWithRetry(req.data());
+    RegionPtr region = context.getTMTContext().getKVStore()->getRegion(region_id);
+    if (!region)
+        throw Exception(fmt::format("No such region: {}", region_id), ErrorCodes::BAD_ARGUMENTS);
+
+    bool unequal_flag = false;
+    DAGProperties properties = getDAGProperties("");
+    std::__1::vector<std::__1::pair<DecodedTiKVKeyPtr, DecodedTiKVKeyPtr>> key_ranges = CoprocessorHandler::GenCopKeyRange(req.ranges());
```

Comment:
To satisfy my curiosity, Could you tell me why use the "std::__1" namespace?

Reply:
Ahoh, they were introduced by CLion's extracting method functionality, I'll change it.

Reply:
Get.